### DISCUSSION TITLE
Learning optimal basis function

### DISCUSSION THREAD

**Discussion by rikhuijzer** [MilesCranmer/PySR#406] [Q&A] [answered]
I'm fitting a model to predict dropout in special forces selection. So far, I have mostly used random forests which outperform linear models. However, Nassim Nicholas Taleb [pointed out](https://twitter.com/nntaleb/status/1691181151451226113) something nice:

<img src="https://github.com/MilesCranmer/PySR/assets/20724914/d987eca1-1d06-47e8-83de-2c7d65288427" width=300>

That is, IQ is not linearly related to fatalities, but only below a certain threshold IQ is linearly related. Therefore, I should probably model it with a ReLU:

<img src="https://github.com/MilesCranmer/PySR/assets/20724914/06e9a8af-1a95-4f23-8c90-d893cf065eb2" width=300>

As also shown in on of the scatter plots in Taleb's [Medium article](https://medium.com/incerto/iq-is-largely-a-pseudoscientific-swindle-f131c101ba39):

![image](https://github.com/MilesCranmer/PySR/assets/20724914/730a5106-e045-4b2b-99e4-a0d08fb5f197)

For my special forces dropout prediction case, it does make a lot of sense. Predicting who will be a great special forces soldier is impossible because it's a power law and those are generally extremely hard to predict. However, you can turn the problem around to predict what's in the way of becoming a soldier. For example, you can be a great person with amazing soldier skills, but if you cannot carry a backpack then the amazing skills don't matter. Similarly, having a great car doesn't matter if one wheel is missing.

Anyway, in my case, I have a $n$ variables (roughly 10) for which I want to feat a ReLU (or something similar) on each variable and then sum them via a logistic (because I'm doing binary classification). So an equation along the lines of 

$y_i = \text{logistic}(b_0 + f(x_1, a_1, b_2, c_2) + f(x_2, a_2, b_2, c_2) + \ ... \ + f(x_n, a_n, b_n, c_n))$

Where $f(x_i, a_i, b_i, c_i)$ denotes some kind of function where SymbolicRegression tries to find the best fitting ReLU with slope $a_i$, location for the angle $b_i$, and direction $c_i$ for feature $x_i$ for all features $i$ in $1:\text{number of features}$.

Would this be possible in SymbolicRegresssion.jl and what would be the best way to do this?

**Comment by MilesCranmer** [comment, marked as answer]
Hi @rikhuijzer,

Thanks for sharing this problem, it sounds very interesting. It is definitely possible, but perhaps not in the exact form you are looking for. The RFC here: https://github.com/MilesCranmer/SymbolicRegression.jl/issues/193 has some ideas for how we could set this up. What do you think? I think it would basically involve making a meta expression type in DynamicExpressions.jl – the meta expression would impose some specific evaluation strategy, and wrap one or more `Node` types for holding the symbolic form of `f`.

There is a way to approach this problem with the current code, but it's a bit brute force. You would do it with a custom objective that would explicitly evaluate the symbolic expression for each of the 10 variables. See https://astroautomata.com/PySR/examples/#9-custom-objectives for an example. This is the Python side of the docs but it's the same API for the Julia version, just replace `full_objective` -> `loss_function`. As seen in the example, you can have a custom functional form or evaluation scheme created within the loss function. One thing you could try to do is have an inner optimization loop over `a1, b1, c1, ...` within the loss function – and for each triplet of parameters, you would copy those into a column of `X`, and then call `eval_tree_array`. However, this seems a bit inefficient because it would need to optimize those variables every time the loss function is called. I guess batching will be really important here to avoid expensive optimizations.

Cheers!
Miles

**Reply by rikhuijzer** [reply]
Thank you @MilesCranmer for your answer. I do appreciate it a lot that you took the time to think about this and see it as an interesting problem. 

For now, I'm not fully understanding your answer, but that is okay I think. I'll fiddle a bit with things and see what works 👍 😄 You have for sure helped me out already. Thank you. Nice animation by the way in the SymbolicRegression.jl README. It helps a lot.

**Reply by MilesCranmer** [reply]
Sounds good! Also did you see my second comment below?

**Comment by MilesCranmer** [comment]
Also - if you are open to not using pure symbolic regression, you could try the deep learning -> symbolic regression approach as mentioned here: https://github.com/MilesCranmer/symbolic_deep_learning.

Here's a full example of how you could solve this with Flux.jl:

```julia
# using CUDA, cuDNN
using Flux
using Fluxperimental: @compact
using Random: seed!
using ProgressBars
using Statistics: mean
using MLJ

seed!(0)

const n_variables = 10
const v_n_variables = Val(n_variables)

n_hidden = 128

##########################################################
## Flexible neural network model to find the optimal basis:
##########################################################
model = @compact(
    f = Chain(
        Dense(4, n_hidden, relu),
        Dense(n_hidden, n_hidden, relu),
        Dense(n_hidden, n_hidden, relu),
        Dense(n_hidden, 1),
    ),
    a = randn(n_variables),
    b = randn(n_variables),
    c = randn(n_variables),
    b0 = randn(),
) do x
    # Assume x: (n_variables, n_rows)
    @assert axes(x, 1) == axes(a, 1)
    nrows = size(x, 2)

    raw_inputs = cat(
        reshape(x, axes(x, 1), 1, axes(x, 2)),
        repeat(a, 1, 1, nrows),
        repeat(b, 1, 1, nrows),
        repeat(c, 1, 1, nrows);
        dims = 2,
    )
    inputs = ntuple(i -> raw_inputs[i, :, :], v_n_variables)

    # inputs: (n_variables, 4, n_rows)
    outputs = sum(f, inputs)

    return sigmoid(outputs .+ b0)[1, :]
end

model = model |> gpu
model(randn(10, 3) |> gpu)  # Should be 3 outputs

##########################################################
## Create dataset:
##########################################################
true_f = (x, a, b, c) -> relu(x * a - b) * c
true_b0 = -0.1
X = randn(Float32, n_variables, 10_000)
unknown_vals = (a = randn(n_variables), b = randn(n_variables), c = randn(n_variables))
Y = [
    sigmoid(
        sum([
            true_f(X[i, j], unknown_vals.a[i], unknown_vals.b[i], unknown_vals.c[i]) for
            i in axes(X, 1)
        ]) + true_b0,
    ) for j in axes(X, 2)
]
loader = Flux.DataLoader((X, Y) |> gpu, batchsize = 64, shuffle = true)

optim = Flux.setup(Flux.Adam(3e-4), model)

##########################################################
## Train network:
##########################################################
losses = []
p = ProgressBar(1:1000)
for i in p
    for (x, y) in loader
        loss, grads = Flux.withgradient(model) do m
            ŷ = m(x)
            Flux.Losses.mse(ŷ, y)
        end
        Flux.update!(optim, model, grads[1])
        push!(losses, loss)
    end
    set_multiline_postfix(p, "Loss = $(mean(losses[end-30:end]))" * " "^40)
end

##########################################################
## Create training data for SymbolicRegression.jl:
##########################################################
sr_X = (a = Float32[], b = Float32[], c = Float32[], x = Float32[])
sr_y = Float32[]
for _ = 1:1000
    i = rand(1:10)
    row = rand(axes(X, 2))
    a = model.variables.a[i]
    b = model.variables.b[i]
    c = model.variables.c[i]

    x = X[i, row]
    y = model.variables.f([x, a, b, c] |> gpu) |> cpu

    push!(sr_X.a, a)
    push!(sr_X.b, b)
    push!(sr_X.c, c)
    push!(sr_X.x, x)
    push!(sr_y, only(y))
end

# Fit the neural net with SymbolicRegression.jl:
sr_regressor = @load SRRegressor
sr_model = sr_regressor(
    unary_operators = [exp, relu, tanh],
    binary_operators = [+, -, *, /],
    niterations = 1000,
)
mach = machine(sr_model, sr_X, sr_y)
fit!(mach)
```

It should hopefully find the basis function. Just keep in mind that it may use the parameters in a different way than the "ground truth" here, because the neural net is not told how to use them - it just uses them to optimize its predictions.

**Reply by MilesCranmer** [reply]
@rikhuijzer in case you didn't see this one^
